src/async_client.rs

```diff
@@ -460,9 +456,9 @@ impl AsyncClient {
 
         debug!("Creating client with persistence: {}, {:?}", ptype, usrptr);
 
-        if !usrptr.is_null() {
-            // TODO: The bridge should return boxed persistence given uptr
-            let persistence = Box::new(ffi::MQTTClient_persistence {
+        let persistence = match usrptr.is_null() {
+            true => ptr::null_mut(),
+            false => &mut ffi::MQTTClient_persistence {
```

Comment:
The `ffi::MQTTClient_persistence` will be dropped at the end of this call, which means the `MQTTAsync` handle will be holding a pointer to freed space. If you Box it up and store it inside of the `AsyncClient` instance than yes, we can change the raw pointer into a Box<ffi::MQTTClient_persistence> and simply pass it as a pointer to the `ffi::MQTTAsync_createWithOptions()` function call and let it deallocate itself naturally when `AsyncClient` is dropped.

Reply:
Yeah, I don't know if I did this one in the safest way possible, but we do have to be careful here. I have a vague memory of my first implementation of this crashing because the persistence object did not live long enough.

The C library is a little inconsistent in how it handles the structures you give it. Often it makes a deep copy, but sometimes it just keeps a pointer, as I believe is the case with persistence.

We could probably Box it rather than using a raw pointer, but we **must** make sure that the persistence object outlives the underlying client.  The C client's `MQTTAsync_destroy()` function calls `close()` on the persistence object. So the persistence object needs to stay in memory until after `MQTTAsync_destroy()` returns.

The persistence object is also optional. So we'd need an Option<Box<..>>, right?

So maybe that's why I went with a raw pointer in the first place: We need to pass a pointer to the C lib; properly manage its lifetime; allow it to be optional; and we're juggling so much unsafe code here, what's a little more :-)

Reply:
Hi all,

Nice to see this moving.

MQTTAsync doesn't hold a pointer to the persistence option but simply memcopy it.
Which does make sense in the big picture.

However this is guaranteed only by the source code, tomorrow it can change at will.

To recap, I believe the code right now will work flawless, at least as long as the underneath c library doesn't change.

However, I don't believe that it can change, once you create the client you cannot change its options simply by modifying the original option object, so the only option is really to memcopy all the option we pass around.

Cheers,
Simone

Reply:
I think I retract what I just said, because afaict the mutable reference that you take on `ffi::MQTTClient_persistence` is immediately invalidated at the end of the `let persistence = match {...} as *mut c_void;` statement, because the `ffi::MQTTClient_persistence` will be dropped right at that point. This is subtle due to the casting of the `&mut` to ` *mut c_void`. Is that not the case?

Reply:
There is one other issue I think that's existed even before this PR.

While the C side will free the `MQTTClient_persistence` after it closes it, it will not free the `context` handle that points to the `ClientPersistence` trait object on the Rust side (this is forgotten via `Box::into_raw`). If we don't maintain a pointer to the `ffi::MQTTClient_persistence` object, we cannot fix this memory leak of dropping the trait object properly.

Reply:
I believe the C library  just copies the pointer table to the persistence callbacks (think shallow copy). So we _do_ need to keep the actual persistence object alive until after the C client is destroyed. I guess it burned me bad enough in the initial implementation that I left myself a note:
https://github.com/eclipse/paho.mqtt.rust/blob/f18ba1be28246be5ea31c900c897054a9eb96dd3/src/async_client.rs#L477-L480

Reply:
The raw persistence pointer from `Box::into_raw` is being re-acquired and dropped in the client's drop function:
https://github.com/eclipse/paho.mqtt.rust/blob/f18ba1be28246be5ea31c900c897054a9eb96dd3/src/async_client.rs#L928

Did I do this correctly?  ...assuming that the C library doesn't call free() on the pointer we give it.

Reply:
Yes you're doing the right thing. The C side only frees the persistence context when `popen` == `pstopen` as seen here: https://github.com/eclipse/paho.mqtt.c/blob/master/src/MQTTPersistence.c#L142

See my updated PR to see what I'm referring to when I am talking about the `ClientPersistence` trait object (the `context`): https://github.com/eclipse/paho.mqtt.rust/pull/11/files#diff-b1c6874b1ea75211abb3fa70a10ab7e5R930

Reply:
If you guys are both convinced that this PR is wrong, please feel free to just close it.

I will try to remove that raw pointer whenever I get the chance.

Reply:
@jnicholls I fixed the first error of raw pointer that you correctly pointed out.

Move the cast inside the call function, so that the `persistence` object goes out of scope after being used and freed later.

Not sure I follow your next point, trying to add some test but I keep finding place where I can fix stuff...